etcdserver/v3_server.go

```diff
@@ -641,3 +646,89 @@ func (s *EtcdServer) processInternalRaftRequest(ctx context.Context, r pb.Intern
 
 // Watchable returns a watchable interface attached to the etcdserver.
 func (s *EtcdServer) Watchable() mvcc.WatchableKV { return s.KV() }
+
+func (s *EtcdServer) linearizableReadLoop() {
+	var (
+		rs       raft.ReadState
+		requests int
+	)
+	ctx := make([]byte, 8)
+	s.readNotifyc = make(chan struct{})
+
+	for {
+		binary.BigEndian.PutUint64(ctx, s.reqIDGen.Next())
+
+		for (len(s.readwaitc) > 0 && requests < 256) || (len(s.readwaitc) == 0 && requests == 0) {
+			select {
+			case <-s.readwaitc:
+				requests++
+			}
+		}
+
+		s.readMu.Lock()
+		var drained bool
+		for !drained {
+			select {
+			// drain all pending requests
+			case <-s.readwaitc:
+			default:
+				drained = true
+			}
+		}
+		requests = 0
+		notifyc := s.readNotifyc
+		s.readNotifyc = make(chan struct{})
+		s.readMu.Unlock()
+
+		if err := s.r.ReadIndex(context.Background(), ctx); err != nil {
+			if err == raft.ErrStopped {
+				return
+			}
+			plog.Errorf("failed to get read index from raft: %v", err)
+			continue
+		}
+
+		select {
+		case rs = <-s.r.readStateC:
+			if !bytes.Equal(rs.RequestCtx, ctx) {
+				plog.Errorf("unexpected read index context value (want %v, got %v)", rs.RequestCtx, ctx)
+				continue
```

Comment:
I don't think `notifyc` can be a bool because it wouldn't be able to broadcast `true` to all waiters. The returned information would need to be wrapped in something like:
```go
type notifier struct {
    // err is the error from the notify set before closing notifyc.
    err error
    notifyc chan struct{}
}
```
where the waiter grabs a pointer to the current `notifier` instead of only getting the current `notifyc`.